File: nodeJS-StarterCode/cityiq.js
// node fetch is required to establish requests
const fetch = require("node-fetch")
// btoa function is required to encode username and password
const btoa = str => new Buffer.from(str).toString('base64')

const pageSize = 1000
const pageNumber = 0
const eventTypes = ['PKIN', 'PKOUT', 'PEDEVT', 'BICYCLE', 'TFEVT', 'HUMIDITY', 'PRESSURE', 'TEMPERATURE', 'METROLOGY', 'ENERGY_TIMESERIES']
const mediaTypes = ['IMAGE', 'VIDEO']
const assetTypes = ['CAMERA', 'ENV_SENSOR', 'NODE', 'EM_SENSOR']
const locationTypes = ['PARKING_ZONE', 'TRAFFIC_LANE', 'WALKWAY']

// requests function formats requests via node.js
function request(url, headers, body) {
  let options = { headers: headers, body: body }
  return fetch(url, options).then(result => {
    if (result.status >= 400) return (result.statusText)
    else return result.text().then(txt => {
      try { return JSON.parse(txt) }
      catch (err) { return txt }
    })
  })
}

// cityiq main function to be called at least once in each demo.js file
module.exports = async function (tenant) {
  console.log('starting requests')
  /* establishing access by getting the client_token with uaa url 
     and username and password. 
     While this request is a basic auth request, the client token will be 
     used for bearer authentication. */
  const bearer = (await request(tenant.uaa, { authorization: 'Basic ' + btoa(tenant.client) }))
  const client_token = bearer.access_token

  const zones = {
    PKIN: tenant.id + "-IE-PARKING",
    PKOUT: tenant.id + "-IE-PARKING",
    PEDEVT: tenant.id + "-IE-PEDESTRIAN",
    TFEVT: tenant.id + "-IE-TRAFFIC",
    BICYCLE: tenant.id + "-IE-BICYCLE",
    HUMIDITY: tenant.id + "-IE-ENVIRONMENTAL",
    PRESSURE: tenant.id + "-IE-ENVIRONMENTAL",
    TEMPERATURE: tenant.id + "-IE-ENVIRONMENTAL",
    METROLOGY: tenant.id + '-IC-METROLOGY',
    ENERGY_TIMESERIES: tenant.id + '-IC-EN_TIMESERIES',
    IMAGE: tenant.id + "-IE-IMAGE",
    VIDEO: tenant.id + "-IE-VIDEO"
  }

  // this function searches assets by their content variables (i.e. assetType, eventTypes, mediaTypes, coordinates etc)
  async function assets(type, id) {
    // id variable can be null

    /** set the variables, query headers and query url will be specified here. 
     * Each request will be made in this function
     * options for eventTypes, mediaTypes and assetTypes are specified here
    */

    let query, headers, queryURL

    /**If statements specify each query for each search parameter. 
     * Default is to search assets by assetUid */
    if (eventTypes.includes(type)) {
      console.log('querying assets by eventTypes')
      query = '/assets/search?bbox=' + tenant.bbox + '&q=eventTypes:' + type + '&page=0&size=' + pageSize
    } else if (mediaTypes.includes(type)) {
      console.log('querying assets by mediaType')
      query = '/assets/search?bbox=' + tenant.bbox + '&q=mediaType:' + type + '&page=0&size=' + pageSize
    } else if (assetTypes.includes(type)) {
      console.log('querying assets by assetType')
      query = '/assets/search?bbox=' + tenant.bbox + '&q=assetType:' + type + '&page=0&size=' + pageSize
    } else if ((id !== undefined) && (['subAssets', 'locations'].includes(type))) {
      console.log('querying for ' + type + ' by uid')
      query = '/assets/' + id + '/' + type
    }
    else {
      console.log('querying by assetUid')
      query = '/assets/' + type
    }

    headers = { authorization: 'Bearer ' + client_token, 'predix-zone-id': zones['TFEVT'] }
    queryURL = tenant.metadataService + query
    console.log('Query URL: ' + queryURL)

    let response = (await request(queryURL, headers))
    // if content is not defined then return the response otherwise return the content.
    return (response.content === undefined) ? response : response.content

  }

  // this function searches locations by their content variables
  async function locations(type, coord = '-90:-180,90:180') {

    let query, headers, queryURL

    if (type === 'bbox') {
      console.log('querying locations by  bbox')
      query = '/locations/search?bbox=' + coord
    } else if (locationTypes.includes(type)) {
      console.log('querying locations by locationType')
      query = '/locations/search?bbox=' + coord + '&q=locationType:' + type
    } else if ((type == 'assets') && (!coord.includes(":"))) {
      console.log('querying for ' + type + ' by uid')
      query = '/locations/' + coord + '/' + type
    } else {
      console.log('querying locations by locationUid')
      query = '/locations/' + type
    }

    headers = { authorization: 'Bearer ' + client_token, 'predix-zone-id': zones['TFEVT'] }
    queryURL = tenant.metadataService + query
    console.log('Query URL: ' + queryURL)

    let response = (await request(queryURL, headers))
    return (response.content === undefined) ? response : response.content
  }

  async function events(id, idType, type, start, stop) {

    let query, headers, queryURL, span


    if (eventTypes.indexOf(type) <= 4) {
      if (idType == 'assetUid') {
        console.log('querying events by asset, eventTypes and time span')
        query = '/assets/' + id + '/events?eventType=' + type
      } else if (idType == 'locationUid') {
        console.log('querying events by locationUid, eventType, and time span')
        query = '/locations/' + id + '/events?eventType=' + type
      } else {
        console.log('querying events by location, eventType and time span')
        query = '/locations/events?eventType=' + type + '&bbox=' + tenant.bbox
      }
    } else if (eventTypes.indexOf(type) >= 5) {
      console.log('querying events by asset, eventTypes and time span')
      query = '/assets/' + id + '/events?eventType=' + type
    } else if (mediaTypes.includes(type)) {
      if (idType == 'assetUid') {
        console.log('querying events by asset, assetType and time span')
        query = '/assets/' + id + '/events?assetType=' + type
      } else if (idType == 'locationUid') {
        console.log('querying events by locationUid, assetType, and time span')
        query = '/locations/' + id + '/events?assetType=' + type
      } else {
        console.log('querying events by location, assetType and time span')
        query = '/locations/events?assetType=' + type + '&bbox=' + coord
      }
    } else {
      console.log('querying assets by externalRefId')
      query = '/assets/' + id + '/events?externalRefId=' + type
    }

    span = '&startTime=' + start + '&endTime=' + ((typeof (stop) == 'undefined') ? '9999999999999' : stop)
    headers = { authorization: 'Bearer ' + client_token, 'predix-zone-id': zones[type] }
    queryURL = tenant.eventService + query + span + '&pageOffset=' + pageNumber + '&pageLimit=' + pageSize
    console.log('Query URL: ' + queryURL)

    let response = (await request(queryURL, headers))
    if (response) {
      console.log("There are a total of " + response.pagination.totalRecords + " events in this timeframe")
      console.log("You are can query for " + (response.pagination.totalPages - 1) + " more pages of " + response.pagination.pageLimit + " events each")

    }
    return (response.content === undefined) ? response : response.content
  }





  return {
    assets: assets,
    locations: locations,
    events: events
  }
}

File: python-StarterCode/cityiq.py
import requests
import base64
import json
import argparse
import time

with open('./myCredentials.json') as json_file:
    credentials = json.load(json_file)
    zones = {
        "PKIN": credentials["id"] + "-IE-PARKING",
        "PKOUT": credentials["id"] + "-IE-PARKING",
        "PEDEVT": credentials["id"] + "-IE-PEDESTRIAN",
        "TFEVT": credentials["id"] + "-IE-TRAFFIC",
        "BICYCLE": credentials["id"] + "-IE-BICYCLE",
        "HUMIDITY": credentials["id"] + "-IE-ENVIRONMENTAL",
        "PRESSURE": credentials["id"] + "-IE-ENVIRONMENTAL",
        "TEMPERATURE": credentials["id"] + "-IE-ENVIRONMENTAL",
        "METROLOGY": credentials["id"] + '-IC-METROLOGY',
        "ENERGY_TIMESERIES": credentials["id"] + '-IC-EN_TIMESERIES',
        "IMAGE": credentials["id"] + "-IE-IMAGE",
        "VIDEO": credentials["id"] + "-IE-VIDEO"
    }


class CityIq(object):
    # # Set up
    def __init__(self, tenantName):
        try:
            self.tenant = credentials
        except KeyError:
            raise Exception("Tenant string specified is not in credentials.py")
        self.token = None
        self.assets = None
        self.locations = None
        self.events = None
        self.bbox = self.tenant["bbox"]

    # # fetching token - only time to use UAA url

    def fetchToken(self):
        headers = {'Authorization': 'Basic ' +
                   (base64.b64encode(bytes(self.tenant["client"], 'ascii'))).decode('ascii')}
        response = requests.request("GET", self.tenant["uaa"], headers=headers)
        self.token = response.json()["access_token"]

    def setToken(self, token):
        self.token = token

    def getToken(self):
        return self.token

    # # bbox set up * optional, can be an input to make manipulating large amounts of nodes easier
    def setBbox(self, bbox):
        self.bbox = bbox

    def getBbox(self):
        return self.bbox

    # # gets the assets or locations by the input parameter filters
    # # path must be "assets" OR "locations"
    # # zone must be "parking", "traffic" "pedestrian" environment" "bicycle" - as seen in credentials.py
    # # filterQ is optional but recommended:
    # #     filterQ is the payload associated to q= in the params.  This can be :
    # #     "eventTypes:PKIN","eventTypes:PKOUT","eventTypes:PEDEVT", etc...
    # #     "assetType:NODE", "assetType:CAMERA", "assetType:ENV_SENSOR", etc...
    # # page is the page# you are querying (starting with 0)
    # # size is the number of assets to apppear on a page
    def fetchMetadata(self, path, filterQ=None, page=0, size=100):
        if self.token is not None:
            query = {
                "bbox": self.bbox,
                "size": str(size),
                "page": str(page)
            }
            if filterQ is not None:
                query["q"] = filterQ
            headers = {'Authorization': 'Bearer ' + self.token,
                       'Predix-Zone-Id': zones["TFEVT"]}
            response = requests.request(
                "GET", self.tenant["metadataService"]+"/"+path+"/search", headers=headers, params=query)

            try:
                # assigns differently depending on if assets or locations
                if path == "assets":
                    self.assets = response.json()["content"]
                else:
                    self.locations = response.json()["content"]

                # informs of more metadata available
                if response.json()["totalElements"] > response.json()["numberOfElements"]:
                    print("More Metadata assets are available for this query.")
                    print("TotalElements: " + str(response.json()[
                          "totalElements"]) + ". Your specified size: " + str(response.json()["size"]))
            except:
                print("Error when fetching metadata")
                print(self.tenant["metadataService"]+"/"+path+"/search")
                print(response, "\n\n")

            return response
        else:
            print("Token required to fetch metadata")

    def getAssets(self):
        return self.assets

    def getLocations(self):
        return self.locations

    # # fetches Events and saves to self.events
    # # path can be "assets" OR "locations"
    # # Uid is the assetUid or locationUid
    # # evType is the eventType as a string i.e. "PKIN"
    # # startTime is time since epoch in milliseconds
    # # endtime is time since epoch in milliseconds
    # # pageLimit is the number of elements response is restricted to
    # # pageOffset is the page you are asking for (default is 0 )

    def fetchEvents(self, path, Uid, evType, startTime, endTime, pageNumber=0, pageSize=1000):
        if self.token is not None:

            # set the query
            query = {
                "pageLimit": str(pageSize),
                "pageOffset": str(pageNumber),
                "eventType": evType,
                "startTime": startTime,
                "endTime": endTime
            }
            headers = {'Authorization': 'Bearer ' +
                       self.token, 'predix-zone-id': zones[evType]}
            response = requests.request(
                "GET", self.tenant["eventService"]+"/"+path+"/"+Uid+"/events", headers=headers, params=query)
            try:
                # sets the content to events
                self.res = response.json()
                self.events = self.res["content"]
                details = self.res["pagination"]
                # informs on pageSize exceeded
                if int(details["totalPages"]) > int(details["pageOffset"]):
                    print("There are a total of " +
                          str(details["totalRecords"])+" events in this timeframe")
                    print("You are can query for " +
                          str(details["totalPages"] - 1)+" more pages of "+ str(details["pageLimit"])+" events each")

                return response
            except Exception as e:
                print("fetchEvents Failed")
                print(response)
                print(e)
                return response

    def getEvents(self):
        return self.events

# for testing purposes


def main(args):
    # use for testing, call above functions here to execute from command line
    return

# for testing purposes


def parse_args():
    '''parse args'''
    parser = argparse.ArgumentParser()
    # put your parser.add_arguments(..) here
    return parser.parse_args()


if __name__ == '__main__':
    main(parse_args())
